Add item requests

AlexanderKolnakov opened the pull request.

AlexanderKolnakov committed:
- корректировки по тестам

AlexanderKolnakov committed:
- переработка метода getAllBooking

AlexanderKolnakov committed:
- переработка метода getAllBooking

AlexanderKolnakov committed:
- тесты - начало

AlexanderKolnakov committed:
- тесты - User контроллер и сервер

AlexanderKolnakov committed:
- стались тесты на сервисы

AlexanderKolnakov committed:
- остались тесты на сервисы букинга и ItemRequest

AlexanderKolnakov committed:
- check style

AlexanderKolnakov committed:
- check style

AlexanderKolnakov committed:
- tests coverage

EvgenyKoptubenko commented on src/test/java/ru/practicum/shareit/booking/BookingControllerTest.java:
```diff
@@ -0,0 +1,178 @@
+
+    @Test
+    @SneakyThrows
+    void createBooking_whenBookingRequestValid_thenReturnBooking() {
```
Если есть желание, попробуй сделать параметризованные тесты https://www.baeldung.com/parameterized-tests-junit-5
Это позволит сократить количество кода, но при этом покрытие не уменьшится

EvgenyKoptubenko commented on src/main/java/ru/practicum/shareit/booking/BookingServiceImpl.java:
```diff
@@ -81,70 +79,72 @@ public BookingDto getBooking(Long userId, Long bookingId) {
     }
```
На 78-й строчке используется deprecated функция, что ок сейчас, но в настоящих проектах лучше не делать.

EvgenyKoptubenko commented:
В BookingServiceImpl есть такая функция
private void checkMayUserGetBookingInfo(Long userId, Long bookingId) {
        if (!(bookingRepository.getById(bookingId).getItem().getOwner().equals(userId)
                || bookingRepository.getById(bookingId).getBooker().getId().equals(userId))) {
            throw new EntityNotFoundException("Пользователь с id " + userId + " не может получить информацию " +
                    "о бронировании с id  " + bookingId + " так как не является ее владельцем или арендатором");
        }
    }
Тут два раза идет обращение к базе данных за одним и тем же 
bookingRepository.getById(bookingId)

Если есть желание, попробуй убрать в коде дублирование запросов к базе данных.

EvgenyKoptubenko commented:
Когда добавляешь пользователя второй раз, то приходит отличное сообщение об ошибке
Пользователь с email232фы322ц222ы2@email.com уже зарегистрирован.
Но при этом код ошибки 500.
Все-таки это не ошибка на стороне сервера. Это скорее Conflict ну или Bad Request

EvgenyKoptubenko commented:
Не работает Patch для пользователя

curl --request PATCH \
  --url http://localhost:8080/users/1 \
  --header 'Content-Type: application/json' \
  --header 'X-Sharer-User-Id: 1' \
  --data '{
	"name":"nameIуddуtem",
	"email":"wrong_emai@lmail.com",
  "login":"logiddуxcыцn222s",
	"birthday":"2012-24-23T18:25:43.511Z"
}'

Попробуй отправить обновление пользователя один раз, потом поменять логин и отправить еще раз.
Сервер пришлет ошибку Пользователь с wrong_emai@lmail.com уже зарегистрирован.

EvgenyKoptubenko reviewed (changes requested):
Работа сделана отлично. Есть немного предложений по улучшению приложения

AlexanderKolnakov committed:
- Параметризованные тесты постараюсь сделать при сдаче 16ого спринта.
- Исправлены остальные замечания: замена deprecated методов репозиториев, избавления от повторных запросов к БД, изменения статусов вывода ошибок, корректировка работа метода по обновлению данных о пользователе

AlexanderKolnakov committed:
-git check style

AlexanderKolnakov committed:
-corrected tests

AlexanderKolnakov committed:
-git check style

AlexanderKolnakov commented:
Параметризованные тесты постараюсь успеть сделать к сдаче 16ого спринта. Уже сталкивался с ними

EvgenyKoptubenko reviewed (approved):
Работа сделана хорошо. Спасибо.
